docs/adr/0008-freeing-objects-and-resources.md

```diff
@@ -0,0 +1,109 @@
+    interface.  In this scenario, there's currently no mechanism for the Rust code to release the
+    Kotlin object's resource.
+
+We should come up with a general approach for resource-holding objects and apply it consistently
+across all languages.
+
+
+## Decision Drivers
+
+* If an object does not hold a resource, then we should be able to use the GC to free it.
+* The solution should work for both Rust objects passed to the foreign language and foreign objects
+  passed to Rust.
+
+## Considered Options
+
+### [Option 1] User-defined close method
+
+* Don't automatically implement any methods to close a resource.
+* Advise users to define a `close()` or similar method for resource-holding objects / callback
+  interfaces / trait interfaces and ensure it's called.  This also likely means defining an error
+  variant like `Error::Closed` that gets returned any methods that use the resource after the
+  `close()` method is called.  Alternatively, handle this by panicking or throwing an error.
+* Use the `Cleaner` API on Kotlin to automatically free UniFFI pointers for Kotlin objects.
+  Note that this combines with the previous bullet.  `close()` releases the resource, the Cleaner
+  action frees the heap allocation.
+* Initially, don't implement `Autoclosable` anymore.  If users want this feature, we could
+  consider supporting `Closable` as a standard trait, like we do with `Debug`, `Display`, etc.
+  In this case, `Closable` wouldn't change the Rust code, instead it would implement
+  `Autoclosable` on Kotlin and maybe implement interfaces for other languages as well.
+* Consider adding a Kotlin configuration option to generate a `destroy()` method to free the Rust
+  pointer instead of relying on the Cleaner.  This option would be for users who need to support
+  older Java versions where Cleaner is not available and users who don't trust that Cleaner is
+  implemented correctly.
+
+### [Option 2] Resource and Object as separate types
+
+* Require UDL users to label their interfaces as either a `Resource` or `Object` and proc-macro
+  users to derive either `Resource` or `Object`.
+* Each type would have slightly different handling.  For example on Kotlin `Resource` would work
+  like an `Object` does today and `Object` would work like the `Object` from [1].
+* For other languages, we could either introduce similar distinctions or make both types behave
+  the same.
+
+### [Option 2] Make Object do both
+
+* Kotlin objects keep their current behavior.  They still have a `destroy` method and also
+  implement `AutoCloseable`.
+* In addition, Kotlin objects also use the `Cleaner` API to ensure they are destroyed if the
+  user doesn't manually call it.
+
+## Pros and Cons of the Options
+
+### [Option 1] User-defined resource close method
+
+* Good, because it's the only option that supports callback interfaces (and trait interfaces in the
+  future).  If a foreign object is passed into Rust and that object holds a resource, Rust needs to
```

Comment:
Foreign object holding a resource would be a callback implementation on foreign language side, right? Any resources being held by the callback are user defined, right? So it should be possible to generate a "close" method for the callback that is called when Rust invokes callback with `0` to remove the callback from handle map. User could then implement the appropriate logic to cleanup any resources being held by the callback. Closing the callback seems to be already implemented, only `close` method is lacking in the interface to allow foreign bindings to close further resources.

Reply:
> Foreign object holding a resource would be a callback implementation on foreign language side, right?

I think we want to generalize callbacks into more generic traits though - eg, wrap them in an `Arc<>` such that Rust code can be blissfully unaware whether a trait implementation is foreign or rust implemented.